test/README.md

```diff
@@ -10,4 +10,96 @@ will build a Fedora 38 qcow2 image using the configuration specified in the file
 
-The config map is also used in CI to dynamically generate test builds using the `./test/cases/generate-build-config` scripts.
+The config map is also used in CI to dynamically generate test builds using the `./test/generate-build-config` and `./test/generate-ostree-build-config` scripts.
+
+## Image build tests in GitLab CI
+
+### Summary
+
+Images are built in GitLab CI when a change in an image definition is detected. The config generator scripts generate all the manifests and create a child pipeline with one job for each manifest and builds the image. On successful build, the result is stored in an s3 bucket by manifest ID. Subsequent runs of the generator script check the cache and only build manifests when their ID is not found in the cache.
+
+### Workflow details
+
+#### 1. Generate build config
+
+The first stage of the workflow runs the `./test/generate-build-config` script.
+
+The config generator:
+- Generates all the manifests for a given architecture using the `./cmd/gen-manifests` tool.
+- Downloads the test build cache.
+- Filters out any manifest with an ID that exists in the build cache.
+  - It also filters out any manifest that depends on an ostree commit because these can't be built without an ostree repository to pull from.
+- For each remaining manifest, creates a build job which runs the `./test/cases/build-image.sh` script for a given distro, image type, and config file.
+  - For ostree container image types (`iot-container` and `edge-container`), it also adds a call to the `./tools/ci/push-container.sh` script to push the container to the GitLab registry. The name and tag for each container is `<build name>:<manifest ID>` (see [Definitions](#definitions) below).
+  - If no builds are needed, it generates a `NullConfig`, which is a simple shell runner that exits successfully. This is required because the child pipeline config cannot be empty.
+
+
+#### 2. Dynamic build job
+
+Each build job runs in parallel. For each image that is successfully built, a file is added to the test build cache under the following path:
+```
+<distro>/<arch>/<manifest ID>.json
+```
+
+Each file in the cache stores information relevant to the build,
+in the form
+```json
+{
+  "distro": "<distro>",
+  "arch": "<arch>",
+  "image-type": "<image type>",
+  "config": "<config name>",
+  "manifest-checksum": "<manifest ID>",
+  "obuild-version": "<osbuild version>",
+  "commit": "<commit ID>"
+}
+```
+
+(see [Definitions](#definitions) below)
+
+for example:
+```json
+{
+  "distro": "fedora-38",
+  "arch": "x86_64",
+  "image-type": "qcow2",
+  "config": "all-customizations",
+  "manifest-checksum": "8c0ce3987d78fe6f3307494cd57ceed861de61c3b04786d6a7f570faacbdb5df",
+  "obuild-version": "osbuild 89",
```

Comment:
Just noticed that the version contains `osbuild`. Why is that so? Wouldn't just the actual version be enough?

Reply:
It takes the output of `osbuild --version` directly.  I didn't want to be too smart about it and parse or cut the output since it's just there for troubleshooting purposes (what did we use to build this).  Ideally, we'd have the osbuild commit SHA but I haven't gotten around to that yet (with osbuild pinning).

Reply:
Thanks for the explanation. It makes sense to keep it as is if this is the output from osbuild itself...